### PR TITLE
Server fixes

### PR DESCRIPTION
**Pull request by shahafn**
- Nonce fix in server
- Gas calculations fixes in server & hub

### CODE REVIEW

File: contracts/RelayHub.sol [hunk 1]
```diff
@@ -200,7 +200,7 @@ contract RelayHub is IRelayHub {
         // This transaction must have enough gas to forward the call to the recipient with the requested amount, and not
         // run out of gas later in this function.
         require(
-            gasleft() >= GAS_RESERVE + requiredGas,
+            initialGas >= GAS_RESERVE + requiredGas,
```
Comment (drortirosh) [inline contracts/RelayHub.sol:203]: Yes, look more right. but how much difference did it make?

Comment (shahafn) [reply contracts/RelayHub.sol:203]: Once I fixed the server gas calculation, this is unpassable. Think about it, it required the whole tx gasLimit + GAS_RESERVE to be lower than the remaining gasLimit after some work. Impossible if the server computes the gasLimit the same way it's computed on-chain.

Comment (forshtat) [reply contracts/RelayHub.sol:203]: This is changed in the light of relay server not being able to estimate how much gas to give to a transaction. Watch the hands: in order to pass the ```gasleft() >= GAS_RESERVE + requiredGas``` check, assuming ```gasleft := txGasLimit - gas_used_to_this_point```, the server must set `txGasLimit := GAS_RESERVE + requiredGas + gas_used_to_this_point`.

`gas_used_to_this_point` is unknown. So it is not a useful require the old way.

File: package.json [hunk 2]
```diff
@@ -49,14 +49,15 @@
     "trace": "MODE=trace truffle test",
     "coverage": "yarn generate && ./scripts/run-coverage",
     "clean-coverage": "shx rm -rf coverage coverage.json",
-    "generate": "rm -rf ./build/contracts && truffle compile && typechain --target trufflehotfix './build/contracts/**/*.json'",
+    "generate": "rm -rf ./build/contracts && npx truffle compile && typechain --target trufflehotfix './build/contracts/**/*.json'",
```
Comment (forshtat) [inline package.json:52]: Do you need 'npx'? I was under impression npm knows to look in your node modules for commands

File: src/relayserver/RelayServer.js [hunk 3]
```diff
@@ -9,13 +9,10 @@ const StakeManagerABI = require('../common/interfaces/IStakeManager')
...
 const RelayRequest = require('../common/EIP712/RelayRequest')
 const utils = require('../common/utils')
-/*
-cannot read TS module if executed by node. Use ts-node to run or, better, fix.
-const Environments = require('../relayclient/types/Environments').environments
-const gtxdatanonzero = Environments.constantinople.gtxdatanonzero
- */
-const gtxdatanonzero = 68
+// const Environments = require('../relayclient/types/Environments')
+const gtxdatanonzero = 16 // Environments.defaultEnvironment.gtxdatanonzero
```
Comment (forshtat) [inline src/relayserver/RelayServer.js:13]: My comments were more useful, I believe

File: contracts/RelayHub.sol (hunk 1 above)
Comment (shahafn) [reply contracts/RelayHub.sol:203]: Yes, and `requiredGas` is supposed to already contain `gas_used_to_this_point`

File: src/relayserver/RelayServer.js [hunk 4]
```diff
@@ -146,7 +144,9 @@ class RelayServer extends EventEmitter {
     }
 
     // Check that max nonce is valid
+    // const releaseMutex = await this.nonceMutex.acquire()
     const nonce = await this._pollNonce()
+    // releaseMutex()
```
Comment (forshtat) [inline src/relayserver/RelayServer.js:149]: \-

File: src/relayserver/RelayServer.js [hunk 5]
```diff
@@ -219,12 +211,10 @@ class RelayServer extends EventEmitter {
       throw new Error('canRelay failed in server: ' + canRelayRet.returnValue)
     }
     // Send relayed transaction
-    const method = this.relayHubContract.methods.relayCall(signedData.message, signature, approvalData)
-    const requiredGas = maxPossibleGas + GAS_RESERVE
     debug('maxPossibleGas is', typeof maxPossibleGas, maxPossibleGas)
-    debug('requiredGas is', typeof requiredGas, requiredGas)
+    // debug('requiredGas is', typeof requiredGas, requiredGas)
```
Comment (forshtat) [inline src/relayserver/RelayServer.js:215]: \-

File: test/RelayServer.test.js [hunk 6]
```diff
@@ -623,6 +632,43 @@ contract('RelayServer', function (accounts) {
     })
   })
 
+  describe('nonce sense', async function () {
```
Comment (forshtat) [inline test/RelayServer.test.js:635]: nonce sense? :-)

File: src/relayserver/RelayServer.js (hunk 3 above)
Comment (shahafn) [reply src/relayserver/RelayServer.js:13]: Yes I will put them back on the new PR

File: test/RelayServer.test.js (hunk 6 above)
Comment (shahafn) [reply test/RelayServer.test.js:635]: Oh right :laughing:

File: src/relayserver/RelayServer.js [hunk 7]
```diff
@@ -197,7 +189,7 @@ class RelayServer extends EventEmitter {
     }
 
     const hubOverhead = parseInt(await this.relayHubContract.methods.getHubOverhead().call())
-    const maxPossibleGas = utils.calculateTransactionMaxPossibleGas({
+    const maxPossibleGas = GAS_RESERVE + utils.calculateTransactionMaxPossibleGas({
```
[the remark this answers is not in the document]
Comment (shahafn) [reply src/relayserver/RelayServer.js:192]: It doesn't take into account the GAS_RESERVE parameter.

File: src/relayserver/RelayServer.js [hunk 8]
```diff
@@ -497,6 +488,8 @@ class RelayServer extends EventEmitter {
     debug('gasPrice', gasPrice)
     const gas = (gasLimit && parseInt(gasLimit)) || await method.estimateGas({ from: this.address }) + 21000
     debug('gasLimit', gas)
+    debug('nonceMutex locked?', this.nonceMutex.isLocked())
+    const releaseMutex = await this.nonceMutex.acquire()
```
[the remark this answers is not in the document]
Comment (shahafn) [reply src/relayserver/RelayServer.js:492]: That is correct.

File: src/relayserver/RelayServer.js (hunk 4 above)
Comment (shahafn) [reply src/relayserver/RelayServer.js]: Fixed

File: src/relayserver/RelayServer.js (hunk 5 above)
Comment (shahafn) [reply src/relayserver/RelayServer.js]: Fixed

File: src/relayserver/RelayServer.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (shahafn) [reply src/relayserver/RelayServer.js:493]: I will add it in a new PR with a test

**Pull request merged**